src/Controller/LogoutController.php

```diff
@@ -0,0 +1,157 @@
+
+
+    /**
+     * @param   Configuration|null  $casConfig
+     * @param   Simple|null         $source
+     * @param   SspContainer|null   $container
+     *
+     * @throws \Exception
+     */
+    public function __construct(
+        private readonly Configuration $sspConfig,
+        // Facilitate testing
+        Configuration $casConfig = null,
+        Simple $source = null,
+        SspContainer $container = null,
+    ) {
+        // We are using this work around in order to bypass Symfony's autowiring for cas configuration. Since
+        // the configuration class is the same, it loads the ssp configuration twice. Still, we need the constructor
+        // argument in order to facilitate testin.
+        $this->casConfig = ($casConfig === null || $casConfig === $sspConfig)
+            ? Configuration::getConfig('module_casserver.php') : $casConfig;
+        $this->authSource = $source ?? new Simple($this->casConfig->getValue('authsource'));
+        $this->container = $container ?? new SspContainer();
+
+        /* Instantiate ticket factory */
+        $this->ticketFactory = new TicketFactory($this->casConfig);
+        /* Instantiate ticket store */
+        $ticketStoreConfig = $this->casConfig->getOptionalValue(
+            'ticketstore',
+            ['class' => 'casserver:FileSystemTicketStore'],
+        );
+        $ticketStoreClass = 'SimpleSAML\\Module\\casserver\\Cas\\Ticket\\'
+            . explode(':', $ticketStoreConfig['class'])[1];
+        $this->ticketStore = new $ticketStoreClass($this->casConfig);
+    }
+
+    /**
+     *
+     * @param   Request      $request
+     * @param   string|null  $url
+     *
+     * @return RedirectResponse|null
+     */
+    public function logout(
+        Request $request,
+        #[MapQueryParameter] ?string $url = null,
+    ): RedirectResponse|null {
+        if (!$this->casConfig->getOptionalValue('enable_logout', false)) {
+            $this->handleExceptionThrown('Logout not allowed');
+        }
+
+        // Skip Logout Page configuration
+        $skipLogoutPage = $this->casConfig->getOptionalValue('skip_logout_page', false);
+
+        if ($skipLogoutPage && $url === null) {
+            $this->handleExceptionThrown('Required URL query parameter [url] not provided. (CAS Server)');
+        }
+
+        // Construct the logout redirect url
+        if ($skipLogoutPage) {
+            $logoutRedirectUrl = $url;
+            $params = [];
+        } else {
+            $logoutRedirectUrl = Module::getModuleURL('casserver/loggedOut');
+            $params =  $url === null ? []
+                : ['url' => $url];
+        }
+
+        // Delete the ticket from the session
+        $session = $this->getSession();
+        if ($session !== null) {
+            $this->ticketStore->deleteTicket($session->getSessionId());
+        }
+
+        // Redirect
+        if (!$this->authSource->isAuthenticated()) {
+            $this->container->redirect($logoutRedirectUrl, $params);
```

Comment:
I've seen some controllers in SSP return a RunnableResponse for areas where ssp will do a redirect. That allows the code to be easier to tests since you don't need to worry about it doing a real redirect, allows the return type to be set to a response, and avoids any chance that something calls exit()

Reply:
Done